builders/networkBuilder.go

```diff
@@ -0,0 +1,824 @@
+
+	return &NetworkBuilder{Network: network}, nil
+}
+
+// Save saves the network builder to an IO.Writer
+func (n *NetworkBuilder) Save(writer io.Writer) error {
+	err := json.NewEncoder(writer).Encode(n.Network)
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+// Build returns the network
+func (n *NetworkBuilder) Build() workloads.Network {
+	return n.Network
+}
+
+// TODO ADD NODE ID TO NETWORK?
+
+// WithName sets the ip range to the network
+func (n *NetworkBuilder) WithName(name string) *NetworkBuilder {
+	n.Network.Name = name
+	return n
+}
+
+// WithIPRange sets the ip range to the network
+func (n *NetworkBuilder) WithIPRange(ipRange schema.IPRange) *NetworkBuilder {
+	n.Network.Iprange = ipRange
+	return n
+}
+
+// WithStatsAggregator sets the stats aggregators to the network
+func (n *NetworkBuilder) WithStatsAggregator(aggregators []workloads.StatsAggregator) *NetworkBuilder {
+	n.Network.StatsAggregator = aggregators
+	return n
+}
+
+// WithNetworkResources sets the network resources to the network
+func (n *NetworkBuilder) WithNetworkResources(netResources []workloads.NetworkNetResource) *NetworkBuilder {
+	n.Network.NetworkResources = netResources
+	return n
+}
+
+// AddNode adds a node
+func (n *NetworkBuilder) AddNode(networkSchema string, nodeID string, subnet string, port uint, forceHidden bool) error {
+	n.NodeID = nodeID
+
+	if subnet == "" {
+		return fmt.Errorf("subnet cannot be empty")
+	}
+	ipnet, err := types.ParseIPNet(subnet)
+	if err != nil {
+		return errors.Wrap(err, "invalid subnet")
+	}
+
+	if port == 0 {
+		port, err = n.pickPort()
+		if err != nil {
+			return errors.Wrap(err, "failed to pick wireguard port")
+		}
+	}
+
+	privateKey, err := wgtypes.GeneratePrivateKey()
+	if err != nil {
+		return errors.Wrap(err, "error during wireguard key generation")
+	}
+	sk := privateKey.String()
+
+	pk, err := crypto.KeyFromID(pkg.StrIdentifier(nodeID))
+	if err != nil {
+		return errors.Wrap(err, "failed to parse nodeID")
+	}
+
+	encrypted, err := crypto.Encrypt([]byte(sk), pk)
+	if err != nil {
+		return errors.Wrap(err, "failed to encrypt private key")
+	}
+
+	pubSubnets, err := n.getEndPointAddrs(pkg.StrIdentifier(nodeID))
+	if err != nil {
+		return errors.Wrap(err, "failed to get node public endpoints")
+	}
+	var endpoints []net.IP
+	if !forceHidden {
+		for _, sn := range pubSubnets {
+			endpoints = append(endpoints, sn.IP)
+		}
+	}
+
+	nr := NetResource{
+		NetworkNetResource: workloads.NetworkNetResource{
+			NodeId:                       nodeID,
+			Iprange:                      schema.IPRange{ipnet.IPNet},
+			WireguardListenPort:          int64(port),
+			WireguardPublicKey:           privateKey.PublicKey().String(),
+			WireguardPrivateKeyEncrypted: hex.EncodeToString(encrypted),
+		},
+		PubEndpoints: endpoints,
+	}
+
+	n.NetResources = append(n.NetResources, nr)
+
+	if err = n.generatePeers(); err != nil {
+		return errors.Wrap(err, "failed to generate peers")
+	}
+
+	f, err := os.Open(networkSchema)
```

Comment:
I don't get this. why do we write things on disk at this point ? I don't believe we should do this. The builder should operated fully in memory. And only if the user wants he can serialize it but that should not be a requirements of the API.